File: apps/web/src/api/predictions.ts
import { normalizeTargetPreview, type TargetFetch, type TargetPreview } from "./targets";

export const DEFAULT_APPLIED_THRESHOLD = 0.95;
export const DEFAULT_BATCH_MAX_TARGETS = 50;

export type BatchRejectedCode =
  | "INVALID_ACCESSION"
  | "MISSING_TARGET_EMBEDDING"
  | "UNSUPPORTED_TARGET";

export type BatchRejectedEntry = {
  code: BatchRejectedCode;
  input: string;
  message: string;
  normalizedTargetUniprotId: string;
};

export type BatchSummary = {
  deduped: number;
  rejected: number;
  submitted: number;
  succeeded: number;
};

export type BatchPredictionResponse = {
  cap: number;
  data: {
    peptideEmbeddingsHash: string;
    proteinEmbeddingsHash: string;
  };
  model: {
    hash: string;
    version: string;
  };
  rejected: BatchRejectedEntry[];
  results: PredictionResponse[];
  summary: BatchSummary;
  threshold: number;
};

export type PredictionRow = {
  classification: string;
  classifierScore: number;
  peptideId: string;
  rank: number;
  sequence: string;
  sourceDescription: string;
};

export type PredictionResponse = {
  data: {
    peptideEmbeddingsHash: string;
    proteinEmbeddingsHash: string;
  };
  model: {
    hash: string;
    version: string;
  };
  predictions: PredictionRow[];
  summary: {
    belowThreshold: number;
    highConfidence: number;
    total: number;
  };
  target: TargetPreview;
  threshold: number;
};

type BackendPredictionResponse = {
  data: {
    peptide_embeddings_hash: string;
    protein_embeddings_hash: string;
  };
  model: {
    hash: string;
    version: string;
  };
  predictions: Array<{
    classification: string;
    classifier_score: number;
    peptide_id: string;
    rank: number;
    sequence: string;
    source_description: string;
  }>;
  summary: {
    below_threshold: number;
    high_confidence: number;
    total: number;
  };
  target: Parameters<typeof normalizeTargetPreview>[0];
  threshold: number;
};

type BackendBatchPredictionResponse = {
  cap: number;
  data: {
    peptide_embeddings_hash: string;
    protein_embeddings_hash: string;
  };
  model: {
    hash: string;
    version: string;
  };
  rejected: Array<{
    code: BatchRejectedCode;
    input: string;
    message: string;
    normalized_target_uniprot_id: string;
  }>;
  results: BackendPredictionResponse[];
  summary: {
    deduped: number;
    rejected: number;
    submitted: number;
    succeeded: number;
  };
  threshold: number;
};

type BackendPredictionError = {
  error?: {
    code?: string;
    message?: string;
  };
  state?: string;
};

export class PredictionApiError extends Error {
  constructor(
    message: string,
    readonly status: number,
    readonly state: string,
    readonly code: string | null,
  ) {
    super(message);
    this.name = "PredictionApiError";
  }
}

export async function runPrediction(
  request: { targetUniprotId: string; threshold: number },
  fetcher: TargetFetch = fetch,
  signal?: AbortSignal,
): Promise<PredictionResponse> {
  const response = await fetcher("/api/predictions", {
    body: JSON.stringify({
      target_uniprot_id: request.targetUniprotId,
      threshold: request.threshold,
    }),
    headers: {
      "Content-Type": "application/json",
    },
    method: "POST",
    signal,
  });
  const payload = (await response.json()) as
    | BackendPredictionError
    | BackendPredictionResponse;

  if (!response.ok) {
    const errorPayload = payload as BackendPredictionError;
    throw new PredictionApiError(
      predictionErrorMessage(errorPayload),
      response.status,
      errorPayload.state ?? "prediction-failed",
      errorPayload.error?.code ?? null,
    );
  }

  return normalizePredictionResponse(payload as BackendPredictionResponse);
}

export function normalizePredictionResponse(
  payload: BackendPredictionResponse,
): PredictionResponse {
  return {
    data: {
      peptideEmbeddingsHash: payload.data.peptide_embeddings_hash,
      proteinEmbeddingsHash: payload.data.protein_embeddings_hash,
    },
    model: payload.model,
    predictions: payload.predictions.map((row) => ({
      classification: row.classification,
      classifierScore: row.classifier_score,
      peptideId: row.peptide_id,
      rank: row.rank,
      sequence: row.sequence,
      sourceDescription: row.source_description,
    })),
    summary: {
      belowThreshold: payload.summary.below_threshold,
      highConfidence: payload.summary.high_confidence,
      total: payload.summary.total,
    },
    target: normalizeTargetPreview(payload.target),
    threshold: payload.threshold,
  };
}

export function predictionErrorMessage(payload: BackendPredictionError): string {
  switch (payload.state) {
    case "not-ready":
      return "Prediction backend is not ready. Check model and data configuration.";
    case "unsupported-target":
      return "Target is not available for prediction.";
    case "artifacts-unavailable":
      return "Prediction embeddings are unavailable.";
    case "model-unavailable":
      return "Prediction model is unavailable.";
    case "missing-target-embedding":
      return "Target is missing the embedding required for prediction.";
    case "rate-limited":
      return "Too many prediction requests. Please wait before retrying.";
    case "invalid-request":
    case "prediction-failed":
      return (
        payload.error?.message ??
        "Prediction could not be completed. Please try again."
      );
    default:
      return (
        payload.error?.message ??
        "Prediction could not be completed. Please try again."
      );
  }
}

export async function runBatchPrediction(
  request: { targetUniprotIds: string[]; threshold: number },
  fetcher: TargetFetch = fetch,
  signal?: AbortSignal,
): Promise<BatchPredictionResponse> {
  const response = await fetcher("/api/predictions/batch", {
    body: JSON.stringify({
      target_uniprot_ids: request.targetUniprotIds,
      threshold: request.threshold,
    }),
    headers: {
      "Content-Type": "application/json",
    },
    method: "POST",
    signal,
  });
  const payload = (await response.json()) as
    | BackendPredictionError
    | BackendBatchPredictionResponse;

  if (!response.ok) {
    const errorPayload = payload as BackendPredictionError;
    throw new PredictionApiError(
      batchErrorMessage(errorPayload),
      response.status,
      errorPayload.state ?? "prediction-failed",
      errorPayload.error?.code ?? null,
    );
  }

  return normalizeBatchPredictionResponse(
    payload as BackendBatchPredictionResponse,
  );
}

export function normalizeBatchPredictionResponse(
  payload: BackendBatchPredictionResponse,
): BatchPredictionResponse {
  return {
    cap: payload.cap,
    data: {
      peptideEmbeddingsHash: payload.data.peptide_embeddings_hash,
      proteinEmbeddingsHash: payload.data.protein_embeddings_hash,
    },
    model: payload.model,
    rejected: payload.rejected.map((entry) => ({
      code: entry.code,
      input: entry.input,
      message: entry.message,
      normalizedTargetUniprotId: entry.normalized_target_uniprot_id,
    })),
    results: payload.results.map(normalizePredictionResponse),
    summary: {
      deduped: payload.summary.deduped,
      rejected: payload.summary.rejected,
      submitted: payload.summary.submitted,
      succeeded: payload.summary.succeeded,
    },
    threshold: payload.threshold,
  };
}

export function batchErrorMessage(payload: BackendPredictionError): string {
  if (
    payload.error?.code === "BATCH_TOO_LARGE" &&
    payload.error?.message
  ) {
    return payload.error.message;
  }

  switch (payload.state) {
    case "not-ready":
      return "Batch prediction backend is not ready. Check model and data configuration.";
    case "artifacts-unavailable":
      return "Batch prediction embeddings are unavailable.";
    case "model-unavailable":
      return "Batch prediction model is unavailable.";
    case "rate-limited":
      return "Too many batch prediction requests. Please wait before retrying.";
    case "invalid-request":
    case "prediction-failed":
      return (
        payload.error?.message ??
        "Batch prediction could not be completed. Please try again."
      );
    default:
      return (
        payload.error?.message ??
        "Batch prediction could not be completed. Please try again."
      );
  }
}

File: apps/api/src/proventl_api/api/predictions.py
import os
from datetime import UTC, datetime
from json import JSONDecodeError

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from proventl_api.cache import PredictionResultCache, cache_database_path
from proventl_api.core.config import (
    ArtifactConfig,
    BackendConfig,
    BackendConfigError,
    load_backend_config,
)
from proventl_api.embeddings import (
    EmbeddingArtifacts,
    EmbeddingValidationError,
    load_embedding_artifacts,
)
from proventl_api.model import (
    KerasModelAdapter,
    ModelValidationError,
    load_model_artifact,
)
from proventl_api.predictions import (
    build_prediction_response_from_probabilities,
    predict_probability_array,
)
from proventl_api.targets.registry import (
    UNIPROT_ACCESSION_PATTERN,
    generate_configured_curated_target_registry,
)

router = APIRouter(prefix="/api", tags=["predictions"])
DEFAULT_BATCH_MAX_TARGETS = 50


@router.post("/predictions")
async def run_prediction(request: Request) -> JSONResponse:
    prediction_request, validation_response = await _parse_prediction_request(request)
    if validation_response is not None:
        return validation_response
    target_uniprot_id, threshold = prediction_request

    try:
        config = load_backend_config()
    except BackendConfigError as error:
        return JSONResponse(
            status_code=503,
            content={
                "state": "not-ready",
                "error": {
                    "code": "BACKEND_NOT_READY",
                    "message": "Prediction backend is not ready.",
                    "details": [
                        {
                            "code": issue.code,
                            "setting": issue.setting,
                            "message": issue.message,
                        }
                        for issue in error.issues
                    ],
                },
            },
        )

    if threshold < 0.50 or threshold > 0.99:
        return _invalid_threshold_response()
    if not UNIPROT_ACCESSION_PATTERN.fullmatch(target_uniprot_id):
        return JSONResponse(
            status_code=400,
            content={
                "state": "invalid-request",
                "normalized_target_uniprot_id": target_uniprot_id,
                "error": {
                    "code": "INVALID_ACCESSION",
                    "field": "target_uniprot_id",
                    "message": "Enter a valid UniProt accession.",
                },
            },
        )

    artifact_by_name = {artifact.name: artifact for artifact in config.artifacts}
    registry = generate_configured_curated_target_registry(
        config=config,
        generated_at=datetime.now(UTC),
    )
    if target_uniprot_id not in {
        target["uniprot_id"].strip().upper()
        for target in registry.targets
    }:
        return JSONResponse(
            status_code=404,
            content={
                "state": "unsupported-target",
                "normalized_target_uniprot_id": target_uniprot_id,
                "error": {
                    "code": "UNSUPPORTED_TARGET",
                    "field": "target_uniprot_id",
                    "message": "Target is not available for prediction.",
                },
            },
        )

    try:
        embeddings = load_embedding_artifacts(
            peptide_artifact=artifact_by_name["peptide_embeddings"],
            protein_artifact=artifact_by_name["protein_embeddings"],
        )
    except EmbeddingValidationError as error:
        return JSONResponse(
            status_code=503,
            content={
                "state": "artifacts-unavailable",
                "error": {
                    "code": "EMBEDDINGS_UNAVAILABLE",
                    "message": "Prediction embeddings are unavailable.",
                    "details": error.issues,
                },
            },
        )
    try:
        model = load_model_artifact(
            model_artifact=artifact_by_name["model"],
            model_loader=request.app.state.model_loader,
        )
    except ModelValidationError as error:
        return JSONResponse(
            status_code=503,
            content={
                "state": "model-unavailable",
                "error": {
                    "code": "MODEL_UNAVAILABLE",
                    "message": "Prediction model is unavailable.",
                    "details": error.issues,
                },
            },
        )

    try:
        probabilities = _cached_or_inferred_probabilities(
            config=config,
            artifact_by_name=artifact_by_name,
            embeddings=embeddings,
            model=model,
            target_uniprot_id=target_uniprot_id,
        )
        content = build_prediction_response_from_probabilities(
            targets=registry.targets,
            embeddings=embeddings,
            model=model,
            target_uniprot_id=target_uniprot_id,
            threshold=threshold,
            probabilities=probabilities,
        )
    except KeyError:
        return JSONResponse(
            status_code=500,
            content={
                "state": "missing-target-embedding",
                "normalized_target_uniprot_id": target_uniprot_id,
                "error": {
                    "code": "MISSING_TARGET_EMBEDDING",
                    "field": "target_uniprot_id",
                    "message": (
                        "Target is missing the embedding required for prediction."
                    ),
                },
            },
        )
    except (ModelValidationError, RuntimeError):
        return JSONResponse(
            status_code=500,
            content={
                "state": "prediction-failed",
                "error": {
                    "code": "PREDICTION_FAILED",
                    "message": "Prediction could not be completed.",
                },
            },
        )

    return JSONResponse(
        status_code=200,
        content=content,
    )


@router.post("/predictions/batch")
async def run_batch_prediction(request: Request) -> JSONResponse:
    prediction_request, validation_response = await _parse_batch_prediction_request(
        request
    )
    if validation_response is not None:
        return validation_response
    target_entries, threshold = prediction_request
    submitted_count = len(target_entries)
    target_entries = _deduplicate_batch_entries(target_entries)
    deduped_count = submitted_count - len(target_entries)
    cap = _batch_max_targets()
    if len(target_entries) > cap:
        return JSONResponse(
            status_code=400,
            content={
                "state": "invalid-request",
                "error": {
                    "code": "BATCH_TOO_LARGE",
                    "field": "target_uniprot_ids",
                    "message": (
                        f"Batch contains {len(target_entries)} deduplicated "
                        f"targets, exceeding the cap of {cap}."
                    ),
                    "cap": cap,
                    "submitted": len(target_entries),
                },
            },
        )
    if threshold < 0.50 or threshold > 0.99:
        return _invalid_threshold_response()

    try:
        config = load_backend_config()
    except BackendConfigError as error:
        return JSONResponse(
            status_code=503,
            content={
                "state": "not-ready",
                "error": {
                    "code": "BACKEND_NOT_READY",
                    "message": "Prediction backend is not ready.",
                    "details": [
                        {
                            "code": issue.code,
                            "setting": issue.setting,
                            "message": issue.message,
                        }
                        for issue in error.issues
                    ],
                },
            },
        )

    artifact_by_name = {artifact.name: artifact for artifact in config.artifacts}
    registry = generate_configured_curated_target_registry(
        config=config,
        generated_at=datetime.now(UTC),
    )
    registry_ids = {
        target["uniprot_id"].strip().upper()
        for target in registry.targets
    }

    try:
        embeddings = load_embedding_artifacts(
            peptide_artifact=artifact_by_name["peptide_embeddings"],
            protein_artifact=artifact_by_name["protein_embeddings"],
        )
    except EmbeddingValidationError as error:
        return JSONResponse(
            status_code=503,
            content={
                "state": "artifacts-unavailable",
                "error": {
                    "code": "EMBEDDINGS_UNAVAILABLE",
                    "message": "Prediction embeddings are unavailable.",
                    "details": error.issues,
                },
            },
        )
    try:
        model = load_model_artifact(
            model_artifact=artifact_by_name["model"],
            model_loader=request.app.state.model_loader,
        )
    except ModelValidationError as error:
        return JSONResponse(
            status_code=503,
            content={
                "state": "model-unavailable",
                "error": {
                    "code": "MODEL_UNAVAILABLE",
                    "message": "Prediction model is unavailable.",
                    "details": error.issues,
                },
            },
        )

    results = []
    rejected = []
    for target_entry in target_entries:
        target_uniprot_id = target_entry["normalized_target_uniprot_id"]
        if not UNIPROT_ACCESSION_PATTERN.fullmatch(target_uniprot_id):
            rejected.append(
                _batch_rejection(
                    target_entry,
                    code="INVALID_ACCESSION",
                    message="Enter a valid UniProt accession.",
                )
            )
            continue
        if target_uniprot_id not in registry_ids:
            rejected.append(
                _batch_rejection(
                    target_entry,
                    code="UNSUPPORTED_TARGET",
                    message="Target is not available for prediction.",
                )
            )
            continue
        try:
            probabilities = _cached_or_inferred_probabilities(
                config=config,
                artifact_by_name=artifact_by_name,
                embeddings=embeddings,
                model=model,
                target_uniprot_id=target_uniprot_id,
            )
            results.append(
                build_prediction_response_from_probabilities(
                    targets=registry.targets,
                    embeddings=embeddings,
                    model=model,
                    target_uniprot_id=target_uniprot_id,
                    threshold=threshold,
                    probabilities=probabilities,
                )
            )
        except KeyError:
            rejected.append(
                _batch_rejection(
                    target_entry,
                    code="MISSING_TARGET_EMBEDDING",
                    message=(
                        "Target is missing the embedding required for prediction."
                    ),
                )
            )
        except (ModelValidationError, RuntimeError, ValueError):
            return JSONResponse(
                status_code=500,
                content={
                    "state": "prediction-failed",
                    "error": {
                        "code": "PREDICTION_FAILED",
                        "message": "Prediction could not be completed.",
                    },
                },
            )

    return JSONResponse(
        status_code=200,
        content={
            "threshold": threshold,
            "cap": cap,
            "summary": {
                "submitted": submitted_count,
                "deduped": deduped_count,
                "succeeded": len(results),
                "rejected": len(rejected),
            },
            "results": results,
            "rejected": rejected,
            "model": {
                "version": model.version,
                "hash": model.short_hash,
            },
            "data": {
                "peptide_embeddings_hash": embeddings.peptide_artifact.short_hash,
                "protein_embeddings_hash": embeddings.protein_artifact.short_hash,
            },
        },
    )


def _cached_or_inferred_probabilities(
    *,
    config: BackendConfig,
    artifact_by_name: dict[str, ArtifactConfig],
    embeddings: EmbeddingArtifacts,
    model: KerasModelAdapter,
    target_uniprot_id: str,
) -> tuple[float, ...]:
    prediction_cache = PredictionResultCache(cache_database_path(config.app_data_dir))
    prediction_cache.migrate()
    cache_key = {
        "target_uniprot_id": target_uniprot_id,
        "model_hash": artifact_by_name["model"].full_hash,
        "peptide_embeddings_hash": artifact_by_name["peptide_embeddings"].full_hash,
        "protein_embeddings_hash": artifact_by_name["protein_embeddings"].full_hash,
    }
    cached_entry = prediction_cache.get(**cache_key)
    if cached_entry is not None and len(cached_entry.probabilities) == len(
        embeddings.peptides.metadata
    ):
        return cached_entry.probabilities

    probabilities = predict_probability_array(
        embeddings=embeddings,
        model=model,
        target_uniprot_id=target_uniprot_id,
    )
    prediction_cache.upsert(
        **cache_key,
        probabilities=probabilities,
        created_at=datetime.now(UTC),
    )
    return probabilities


def _deduplicate_batch_entries(
    entries: list[dict[str, str]],
) -> list[dict[str, str]]:
    seen = set()
    deduplicated = []
    for entry in entries:
        normalized_target_id = entry["normalized_target_uniprot_id"]
        if normalized_target_id in seen:
            continue
        seen.add(normalized_target_id)
        deduplicated.append(entry)
    return deduplicated


def _batch_rejection(
    target_entry: dict[str, str],
    *,
    code: str,
    message: str,
) -> dict[str, str]:
    return {
        "input": target_entry["input"],
        "normalized_target_uniprot_id": target_entry["normalized_target_uniprot_id"],
        "code": code,
        "message": message,
    }


def _invalid_threshold_response() -> JSONResponse:
    return JSONResponse(
        status_code=400,
        content={
            "state": "invalid-request",
            "error": {
                "code": "INVALID_THRESHOLD",
                "field": "threshold",
                "message": "Threshold must be between 0.50 and 0.99.",
            },
        },
    )


async def _parse_prediction_request(
    request: Request,
) -> tuple[tuple[str, float] | None, JSONResponse | None]:
    try:
        payload = await request.json()
    except JSONDecodeError:
        return None, _validation_error(
            [
                {
                    "code": "MALFORMED_JSON",
                    "message": "Request body must be valid JSON.",
                }
            ]
        )

    if not isinstance(payload, dict):
        return None, _validation_error(
            [
                {
                    "code": "INVALID_BODY",
                    "message": "Prediction request body must be a JSON object.",
                }
            ]
        )

    missing_field_issues = [
        {
            "code": "MISSING_FIELD",
            "field": field,
            "message": f"{field} is required.",
        }
        for field in ("target_uniprot_id", "threshold")
        if field not in payload
    ]
    if missing_field_issues:
        return None, _validation_error(missing_field_issues)

    target_uniprot_id = str(payload["target_uniprot_id"]).strip().upper()
    try:
        threshold = float(payload["threshold"])
    except (TypeError, ValueError):
        return None, _validation_error(
            [
                {
                    "code": "INVALID_FIELD",
                    "field": "threshold",
                    "message": "threshold must be a number.",
                }
            ]
        )

    return (target_uniprot_id, threshold), None


async def _parse_batch_prediction_request(
    request: Request,
) -> tuple[tuple[list[dict[str, str]], float] | None, JSONResponse | None]:
    try:
        payload = await request.json()
    except JSONDecodeError:
        return None, _validation_error(
            [
                {
                    "code": "MALFORMED_JSON",
                    "message": "Request body must be valid JSON.",
                }
            ]
        )

    if not isinstance(payload, dict):
        return None, _validation_error(
            [
                {
                    "code": "INVALID_BODY",
                    "message": "Prediction request body must be a JSON object.",
                }
            ]
        )

    missing_field_issues = [
        {
            "code": "MISSING_FIELD",
            "field": field,
            "message": f"{field} is required.",
        }
        for field in ("target_uniprot_ids", "threshold")
        if field not in payload
    ]
    if missing_field_issues:
        return None, _validation_error(missing_field_issues)

    raw_target_ids = payload["target_uniprot_ids"]
    if not isinstance(raw_target_ids, list) or not raw_target_ids:
        return None, _validation_error(
            [
                {
                    "code": "VALIDATION_ERROR",
                    "field": "target_uniprot_ids",
                    "message": "target_uniprot_ids must be a non-empty array.",
                }
            ]
        )

    target_entries = [
        {
            "input": str(target_id),
            "normalized_target_uniprot_id": str(target_id).strip().upper(),
        }
        for target_id in raw_target_ids
    ]
    try:
        threshold = float(payload["threshold"])
    except (TypeError, ValueError):
        return None, _validation_error(
            [
                {
                    "code": "INVALID_FIELD",
                    "field": "threshold",
                    "message": "threshold must be a number.",
                }
            ]
        )

    return (target_entries, threshold), None


def _batch_max_targets() -> int:
    try:
        parsed = int(os.environ.get("PROVENTL_BATCH_MAX_TARGETS", ""))
    except ValueError:
        return DEFAULT_BATCH_MAX_TARGETS
    if parsed <= 0:
        return DEFAULT_BATCH_MAX_TARGETS
    return parsed


def _validation_error(details: list[dict[str, object]]) -> JSONResponse:
    return JSONResponse(
        status_code=400,
        content={
            "state": "invalid-request",
            "error": {
                "code": "VALIDATION_ERROR",
                "message": "Prediction request is invalid.",
                "details": details,
            },
        },
    )
